### PR TITLE
add partial_concat op in contrib

### PR DESCRIPTION
**Pull request by ForFishes**
Add partial_concat op implementation and the unit test codes of it in contrib. Different from concat op, this op can concat the vars by specifying the initial position(start_index) and length(length). This op is used to concat 'vars' up in part.  This OP exists in contrib, which means that it is not shown to the public.

### CODE REVIEW

File: paddle/fluid/operators/partial_concat_op.cu [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+      else if (length <= sm_count * 128)
+        tile_size = 256;
+      grids = dim3(CEIL_DIV(length, tile_size), 1, 1);
+      blocks = dim3(tile_size, 1, 1);
+    };
+
+    auto place = ctx.GetPlace();
+    T *out_data = out->mutable_data<T>(place);
+
+    std::vector<const T *> in_data;
```
Comment (gavin1332) [inline paddle/fluid/operators/partial_concat_op.cu:116]: do not use dynamic vector, use size-fixed array: `T** in_data[in_num];`

File: paddle/fluid/operators/partial_concat_op.cu [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    dim3 blocks;
+    auto ComputeKernelParameter = [&](size_t length) {
+      if (length >= max_threads)
+        tile_size = 1024;
+      else if (length < max_threads && length > sm_count * 128)
+        tile_size = 512;
+      else if (length <= sm_count * 128)
+        tile_size = 256;
+      grids = dim3(CEIL_DIV(length, tile_size), 1, 1);
+      blocks = dim3(tile_size, 1, 1);
```
Comment (gavin1332) [inline paddle/fluid/operators/partial_concat_op.cu]: as the grids and blocks have only one meaningful dimension, why not call cuda kernel by `call_kernel<<<CEIL_DIV(length, tile_size), tile_size, 0, stream>>>` directly?

**Review by gavin1332 — COMMENTED** [review]
the kernel implementation is OK

File: python/paddle/fluid/contrib/layers/nn.py [hunk 3]
```diff
@@ -24,17 +24,14 @@
...
-    'var_conv_2d',
-    'match_matrix_tensor',
-    'tree_conv',
-    'fused_embedding_seq_pool',
-    'multiclass_nms2',
-    'search_pyramid_hash',
-    'shuffle_batch',
+    'fused_elemwise_activation', 'sequence_topk_avg_pooling', 'var_conv_2d',
+    'match_matrix_tensor', 'tree_conv', 'fused_embedding_seq_pool',
+    'multiclass_nms2', 'search_pyramid_hash', 'shuffle_batch', 'partial_concat'
```
Comment (hutuxian) [inline python/paddle/fluid/contrib/layers/nn.py:34]: why change this format?

File: paddle/fluid/operators/partial_concat_op.cu (hunk 1 above)
Comment (ForFishes) [reply paddle/fluid/operators/partial_concat_op.cu:116]: Done.

File: paddle/fluid/operators/partial_concat_op.cu (hunk 2 above)
Comment (ForFishes) [reply paddle/fluid/operators/partial_concat_op.cu]: OK，I will fix it.

File: python/paddle/fluid/contrib/layers/nn.py (hunk 3 above)
Comment (ForFishes) [reply python/paddle/fluid/contrib/layers/nn.py:34]: It was changed by the `pre-commit` automatically.

**Review by gavin1332 — DISMISSED** [review]
LGTM

**Comment by lanxianghit** [comment]
没有文档预览图，没有中文文档，代码示例不够丰富，需要给出含有实际数据的例子并给出运算结果，start和length取值的特殊情况没有说明，比如越界情况下的行为，负数的行为等

File: python/paddle/fluid/contrib/layers/nn.py [hunk 4]
```diff
@@ -808,3 +805,46 @@ def shuffle_batch(x, seed=None):
                  'SeedOut': seed},
         attrs=op_attrs)
     return out
+
+
+def partial_concat(input, start_index=0, length=-1):
```
Comment (lanxianghit) [inline python/paddle/fluid/contrib/layers/nn.py:810]: python接口定义需要TPM review，包括参数设置，例如使用start+length还是使用start+end

File: paddle/fluid/operators/partial_concat_op.cc [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    AddAttr<int>("start_index",
+                 "The start index of each instance for concatenation.")
+        .SetDefault(0);
+    AddAttr<int>("length",
+                 "The length of each instance for concatenation."
+                 " Negative values for all elements after start_index")
+        .SetDefault(-1);
+    AddComment(R"DOC(
+Partial Concat Operator.
+Partial Concatenate the input tensors along the 2nd dimension.
```
Comment (XiaoguangHu01) [inline paddle/fluid/operators/partial_concat_op.cc:156]: 文档可以说明下只支持2维的Tensor，并且只能沿着第2维进行slice和concat。

File: paddle/fluid/operators/partial_concat_op.cc [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+    return framework::OpKernelType(OperatorWithKernel::IndicateVarDataType(
+                                       ctx, framework::GradVarName("Out")),
+                                   ctx.device_context());
+  }
+};
+
+class PartialConcatOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X", "Input tensors of concat operator.").AsDuplicable();
```
Comment (XiaoguangHu01) [inline paddle/fluid/operators/partial_concat_op.cc:145]: 说明下输入X只能支持2维的Tensor，是否支持LodTensor?

File: python/paddle/fluid/contrib/layers/nn.py [hunk 7]
```diff
@@ -808,3 +805,63 @@ def shuffle_batch(x, seed=None):
...
+        start_index(int32): The start index of each instance for partial concatenation.
+            Default is 0.
+        length(int32): The length of each instance for partial concatenation. Default is -1.
+            Negative values for all elements after start_index.
+    Returns:
+        Variable: A Tensor with the same data type as input's.
+    Examples:
+        .. code-block:: python
+            import paddle.fluid as fluid
+            x = fluid.layers.data(name="x", shape=[3], dtype="float32")
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/contrib/layers/nn.py]: 建议用fluid.data API

File: python/paddle/fluid/contrib/layers/nn.py [hunk 8]
```diff
@@ -808,3 +805,63 @@ def shuffle_batch(x, seed=None):
                  'SeedOut': seed},
         attrs=op_attrs)
     return out
+
+
+def partial_concat(input, start_index=0, length=-1):
+    """
+    **Partial Concat**
+    This OP concatenates the inputs according to the start index and length. This
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/contrib/layers/nn.py:813]: 可以说明下只支持2维的Tensor输入，只支持第2维度的slice和concat

File: python/paddle/fluid/tests/unittests/test_layers.py [hunk 9]
```diff
@@ -2912,6 +2912,16 @@ def test_unfold(self):
             out = layers.unfold(x, [3, 3], 1, 1, 1)
             return (out)
 
+    def test_partial_concat(self):
+        with self.static_graph():
+            x = layers.data(name="x", shape=[3], dtype="float32")
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/tests/unittests/test_layers.py]: 建议用fluid.data API

Comment (ForFishes) [reply python/paddle/fluid/tests/unittests/test_layers.py]: Done.

File: python/paddle/fluid/contrib/layers/nn.py (hunk 8 above)
Comment (ForFishes) [reply python/paddle/fluid/contrib/layers/nn.py:813]: Ok, i will fix it.

File: python/paddle/fluid/contrib/layers/nn.py (hunk 7 above)
Comment (ForFishes) [reply python/paddle/fluid/contrib/layers/nn.py]: OK, done.

File: paddle/fluid/operators/partial_concat_op.cc (hunk 6 above)
Comment (ForFishes) [reply paddle/fluid/operators/partial_concat_op.cc:145]: ok.done

File: paddle/fluid/operators/partial_concat_op.cc (hunk 5 above)
Comment (ForFishes) [reply paddle/fluid/operators/partial_concat_op.cc:156]: ok, i will fix it

File: python/paddle/fluid/contrib/layers/nn.py (hunk 4 above)
Comment (ForFishes) [reply python/paddle/fluid/contrib/layers/nn.py:810]: ok, done.

**Review by lanxianghit — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]

**Review by Boyan-Liu — APPROVED** [review]

File: python/paddle/fluid/contrib/layers/nn.py (hunk 4 above)
Comment (Boyan-Liu) [reply python/paddle/fluid/contrib/layers/nn.py:810]: > python接口定义需要TPM review，包括参数设置，例如使用start+length还是使用start+end

对应业务需求：蜂巢广告的预测模型，start+end用户可能更方便，不用计算长度，后续可根据排期加入。

**Pull request merged**